compiler/liveness.go

```diff
@@ -1,54 +1,264 @@
 // Value liveness analysis & register allocation.
-
+//
+// Liveness implementation inspired by https://hal.inria.fr/inria-00558509v2
+//
+// - A variable a is live-in at node n if it is used at n (a ∈ PhiUses(B)),
+// or if there is a path from n to a node that uses a that does not contain a
+// re-definition of a.
+//
+// - A variable a is live-out at node n if it is live-in at one
+// of the successors of n.
 package compiler
 
-// FIXME: The current RegAlloc is based on wasm stack info and we probably
-// want a real one (in addition to this) with liveness analysis.
-// Returns the total number of registers used.
-func (c *SSAFunctionCompiler) RegAlloc() int {
-	regID := TyValueID(1)
-	valueRelocs := make(map[TyValueID]TyValueID)
-	for _, values := range c.StackValueSets {
-		for _, v := range values {
-			valueRelocs[v] = regID
-		}
-		regID++
-	}
-	for i := range c.Code {
-		ins := &c.Code[i]
-
-		if ins.Target != 0 {
-			if reg, ok := valueRelocs[ins.Target]; ok {
-				ins.Target = reg
-			} else {
-				panic("Register not found for target")
-			}
+import (
+	"github.com/go-interpreter/wagon/wasm"
+	"github.com/golang-collections/collections/stack"
```

Comment:
The stack datastructure should probably be moved in this repo.

Reply:
Yep, typesafe stack will be a good addition.